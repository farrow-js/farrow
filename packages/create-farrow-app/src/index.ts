import fs from 'fs'
import path from 'path'
import minimist from 'minimist'
import { prompt } from 'enquirer'
import { green, blue, stripColors, cyan } from 'kolorist'

const argv = minimist(process.argv.slice(2))

const cwd = process.cwd()

// prettier-ignore
const TEMPLATES = [
  blue('farrow-only'),
  green('farrow-vite-react'),
  cyan('farrow-next-react')
]

const renameFiles: Record<string, string> = {
  _gitignore: '.gitignore',
}

type ValidTemplate = {
  isValid: true
  template: string
}

type InvalidTemplate = {
  isValid: false
  isEmpty?: boolean
  message: string
}

type TemplateStatus = ValidTemplate | InvalidTemplate

const validateTemplate = (template: unknown): TemplateStatus => {
  if (typeof template !== 'string') {
    return {
      isValid: false,
      isEmpty: true,
      message: `Expected template to be a string, instead of ${template}`,
    }
  }

  const availableTemplates = TEMPLATES.map(stripColors)

  if (availableTemplates.includes(template)) {
    return {
      isValid: true,
      template,
    }
  }

  return {
    isValid: false,
    message: `Unexpected template: ${template}, the available templates are ${availableTemplates.join(', ')}`,
  }
}

async function init() {
  let targetDir = argv._[0]

  if (!targetDir) {
    const { name } = await prompt<{ name: string }>({
      type: 'input',
      name: 'name',
      message: `Project name:`,
      initial: 'farrow-project',
    })
    targetDir = name
  }

  const root = path.join(cwd, targetDir)
  console.log(`\nScaffolding project in ${root}...`)

  if (!fs.existsSync(root)) {
    fs.mkdirSync(root, { recursive: true })
  } else {
    const existing = fs.readdirSync(root)
    if (existing.length) {
      const { yes } = await prompt<{ yes: boolean }>({
        type: 'confirm',
        name: 'yes',
        initial: 'Y',
        message: `Target directory ${targetDir} is not empty.\n` + `Remove existing files and continue?`,
      })
      if (yes) {
        emptyDir(root)
      } else {
        return
      }
    }
  }

  // determine template
  let template = argv.t || argv.template

  const templateStatus = validateTemplate(template)

  if (!templateStatus.isValid) {
    const message = templateStatus.isEmpty
      ? 'Select a template:'
      : `${template} isn't a valid template. Please choose from below:`

    const { t } = await prompt<{ t: string }>({
      type: 'select',
      name: 't',
      message,
      choices: TEMPLATES,
    })
    template = stripColors(t)
  }

  const templateDir = path.join(__dirname, `../templates/${template}`)

  const write = (file: string, content?: string) => {
    const targetPath = renameFiles[file] ? path.join(root, renameFiles[file]) : path.join(root, file)
    if (content) {
      fs.writeFileSync(targetPath, content)
    } else {
      copy(path.join(templateDir, file), targetPath)
    }
  }

  const files = fs.readdirSync(templateDir).filter((file) => {
    return file !== 'package.json' && !file.includes('node_modules')
  })

  for (const file of files) {
    write(file)
  }

  const pkg = require(path.join(templateDir, `package.json`))

  pkg.name = path
    .basename(root)
    // #2360 ensure package.json name is valid
    .trim()
    .replace(/\s+/g, '-')
    .replace(/^[._]/, '')
    .replace(/[~)('!*]+/g, '-')

  write('package.json', JSON.stringify(pkg, null, 2))

  const pkgManager = /yarn/.test(process.env.npm_execpath ?? '') ? 'yarn' : 'npm'

  console.log(`\nDone. Now run:\n`)
  if (root !== cwd) {
    console.log(`  cd ${path.relative(cwd, root)}`)
  }
  console.log(`  ${pkgManager === 'yarn' ? `yarn` : `npm install`}`)
  console.log(`  ${pkgManager === 'yarn' ? `yarn dev` : `npm run dev`}`)
}

function copy(src: string, dest: string) {
  const stat = fs.statSync(src)
  if (stat.isDirectory()) {
    copyDir(src, dest)
  } else {
    fs.copyFileSync(src, dest)
  }
}

function copyDir(srcDir: string, destDir: string) {
  fs.mkdirSync(destDir, { recursive: true })
  for (const file of fs.readdirSync(srcDir)) {
    const srcFile = path.resolve(srcDir, file)
    const destFile = path.resolve(destDir, file)
    copy(srcFile, destFile)
  }
}

function emptyDir(dir: string) {
  if (!fs.existsSync(dir)) {
    return
  }
  for (const file of fs.readdirSync(dir)) {
    const abs = path.resolve(dir, file)
    // baseline is Node 12 so can't use rmSync :(
    if (fs.lstatSync(abs).isDirectory()) {
      emptyDir(abs)
      fs.rmdirSync(abs)
    } else {
      fs.unlinkSync(abs)
    }
  }
}

init().catch((e) => {
  console.error(e)
})
