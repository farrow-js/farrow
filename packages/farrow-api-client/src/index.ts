export * from './farrow-api-client';