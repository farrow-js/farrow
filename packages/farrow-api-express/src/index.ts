export * from './apiRouter'
export * from './context'
