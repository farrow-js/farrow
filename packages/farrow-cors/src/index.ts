import { Response, useReq, useRes, MaybeAsyncResponse } from 'farrow-http'
import type { IncomingMessage } from 'http'
import type { Middleware } from 'farrow-pipeline'

import Cors, { CorsOptions, CorsOptionsDelegate, CorsRequest } from 'cors'
import { promisify } from 'util'

export { CorsOptions, CorsOptionsDelegate, CorsRequest }

export const cors = (
  options?: CorsOptions | CorsOptionsDelegate<IncomingMessage>,
): Middleware<any, MaybeAsyncResponse> => {
  const cors = promisify(
    Cors(
      typeof options === 'function'
        ? options
        : {
            ...options,
            preflightContinue: true,
          },
    ),
  )

  return async (request, next) => {
    const req = useReq()
    const res = useRes()

    try {
      await cors(req, res)
      if (req.method?.toLowerCase() === 'options') {
        if (typeof options === 'object') {
          if (options.preflightContinue) {
            return next(request)
          }
        }
        return Response.status(204).string('')
      }
      return next(request)
    } catch (error: any) {
      return Response.status(500).text(error.message)
    }
  }
}
