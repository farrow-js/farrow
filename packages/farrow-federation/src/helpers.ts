import nodeFetch from 'node-fetch'
import type { FormatApi } from 'farrow-api/dist/toJSON'
import type { FormatType, FormatField, FormatFields } from 'farrow-schema/formatter'
import type { Fetch, Fetcher } from './federation'

export const createFetcher = (fetch: Fetch = nodeFetch as any): Fetcher => {
  return async (request) => {
    const { url, calling, options: init } = request

    const options: RequestInit = {
      method: 'POST',
      credentials: 'include',
      ...init,
      headers: {
        'Content-Type': 'application/json',
        ...init?.headers,
      },
      body: JSON.stringify(calling),
    }
    const response = await fetch(url, options)

    return response.json()
  }
}

export const updateType = (type: FormatType, base: number): FormatType => {
  switch (type.type) {
    case 'Scalar': {
      return type
    }
    case 'Object': {
      return {
        ...type,
        fields: updateFields(type.fields, base),
      }
    }
    case 'Union': {
      const itemTypes = type.itemTypes.map((type) => {
        const typeId = type.typeId + base
        return {
          ...type,
          typeId,
          $ref: `#/types/${typeId}`,
        }
      })

      return {
        ...type,
        itemTypes,
      }
    }
    case 'Struct': {
      return {
        ...type,
        fields: updateFields(type.fields, base),
      }
    }
    case 'Record': {
      const valueTypeId = type.valueTypeId + base
      return {
        ...type,
        valueTypeId,
        $ref: `#/types/${valueTypeId}`,
      }
    }
    case 'List': {
      const itemTypeId = type.itemTypeId + base
      return {
        ...type,
        itemTypeId,
        $ref: `#/types/${itemTypeId}`,
      }
    }
    case 'Literal': {
      return type
    }
    case 'Nullable': {
      const itemTypeId = type.itemTypeId + base
      return {
        ...type,
        itemTypeId,
        $ref: `#/types/${itemTypeId}`,
      }
    }
    case 'Intersect': {
      const itemTypes = type.itemTypes.map((type) => {
        const typeId = type.typeId + base
        return {
          ...type,
          typeId,
          $ref: `#/types/${typeId}`,
        }
      })

      return {
        ...type,
        itemTypes,
      }
    }
    case 'Tuple': {
      const itemTypes = type.itemTypes.map((type) => {
        const typeId = type.typeId + base
        return {
          ...type,
          typeId,
          $ref: `#/types/${typeId}`,
        }
      })

      return {
        ...type,
        itemTypes,
      }
    }
    case 'Strict': {
      const itemTypeId = type.itemTypeId + base
      return {
        ...type,
        itemTypeId,
        $ref: `#/types/${itemTypeId}`,
      }
    }
    case 'NonStrict': {
      const itemTypeId = type.itemTypeId + base
      return {
        ...type,
        itemTypeId,
        $ref: `#/types/${itemTypeId}`,
      }
    }
    case 'ReadOnly': {
      const itemTypeId = type.itemTypeId + base
      return {
        ...type,
        itemTypeId,
        $ref: `#/types/${itemTypeId}`,
      }
    }
    case 'ReadOnlyDeep': {
      const itemTypeId = type.itemTypeId + base
      return {
        ...type,
        itemTypeId,
        $ref: `#/types/${itemTypeId}`,
      }
    }
    default: {
      throw new Error(`Unknown type: ${(type as FormatType).type} of FormatType`)
    }
  }
}

export const updateApi = (api: FormatApi, base: number): FormatApi => {
  return {
    ...api,
    input: updateField(api.input, base),
    output: updateField(api.output, base),
  }
}

export const updateFields = (fields: FormatFields, base: number): FormatFields => {
  const newFields: FormatFields = {}

  for (const key in fields) {
    newFields[key] = updateField(fields[key], base)
  }

  return newFields
}

export const updateField = (type: FormatField, base: number): FormatField => {
  const typeId = type.typeId + base

  return {
    ...type,
    typeId,
    $ref: `#/types/${typeId}`,
  }
}
