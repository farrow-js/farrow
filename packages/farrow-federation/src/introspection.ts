import nodeFetch from 'node-fetch'
import type { FormatResult } from 'farrow-api/dist/toJSON'
import { Result, Ok, Err } from './result'
import type { Fetch } from './federation'

export type IntrospectionResult = Result<FormatResult>

export const getIntrospection = (src: string, fetch: Fetch = nodeFetch as any): Promise<IntrospectionResult> => {
  const url = `${src}/__introspection__`
  return fetch(url)
    .then((response) => {
      return response.json() as Promise<FormatResult>
    })
    .then((result) => {
      return Ok(result)
    })
    .catch((err) => {
      return Err(err.stack || err.message)
    })
}
