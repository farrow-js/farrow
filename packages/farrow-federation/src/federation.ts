import { Router, Response } from 'farrow-http'
import nodeFetch from 'node-fetch'
import {
  SingleCalling,
  BatchCalling,
  ApiSingleResponse,
  ApiErrorResponse,
  ApiBatchSuccessResponse,
} from 'farrow-api-server'
import { getFederationInfo } from './info'
import { createFetcher } from './helpers'

export type ApiService = {
  url: string
  namespace: string
}

export type ApiServices = ApiService[]

export type ApiEntryMap = Map<string, string>

export type Fetch = typeof fetch

export type ApiSingleRequest = {
  url: string
  calling: SingleCalling
  options?: RequestInit
}

export type ApiBatchRequest = {
  url: string
  calling: BatchCalling
  options?: RequestInit
}
export type Fetcher = ((request: ApiSingleRequest) => Promise<ApiSingleResponse>) &
  ((request: ApiBatchRequest) => Promise<ApiBatchSuccessResponse>)

export type FederationOptions = {
  fetch?: Fetch
  fetcher?: Fetcher
  polling?: boolean
  pollingInterval?: number
  errorStack?: boolean
  strict?: boolean
}

const defaultOptions = {
  fetch: nodeFetch as any,
  polling: false,
  pollingInterval: 3000,
  strict: true,
}

const UNKNOWN_REQUEST_MESSAGE = 'Unknown structure of request'

export const createFederationServices = async (services: ApiServices, customOptions: FederationOptions = {}) => {
  const isNotProduction = process.env.NODE_ENV !== 'production'
  const options: Required<FederationOptions> = {
    ...defaultOptions,
    errorStack: isNotProduction,
    fetcher: createFetcher(customOptions?.fetch || defaultOptions.fetch),
    ...customOptions,
  }

  const router = Router()

  let info = await getFederationInfo(services, options)
  const polling = async () => {
    info = await getFederationInfo(services, options)
  }

  const getApiEntry = (namespace: string): string | undefined => {
    return info.entryMap.get(namespace)
  }

  const handleCalling = (
    calling: SingleCalling,
    init: RequestInit = {},
  ): Promise<ApiSingleResponse> | ApiSingleResponse => {
    const apiEntry = getApiEntry(calling.path[0])

    if (!apiEntry) {
      const message = `The target API was not found with the path: [${calling.path.join(', ')}]`
      return ApiErrorResponse(message)
    }

    const wrapCalling: SingleCalling = {
      ...calling,
      path: calling.path.slice(1),
    }

    const request: ApiSingleRequest = {
      url: apiEntry,
      calling: wrapCalling,
      options: init,
    }

    return options.fetcher(request)
  }

  router.use(async (request, next) => {
    if (request.method?.toLowerCase() !== 'post') {
      return next()
    }

    const init: RequestInit = {
      headers: request.headers,
    }

    if (request.body?.type === 'Batch') {
      // batch calling
      const callings = request.body!.callings

      if (Array.isArray(callings)) {
        const result = await Promise.all(callings.map((calling) => handleCalling(calling, init)))
        return Response.json(ApiBatchSuccessResponse(result))
      }

      return Response.json(ApiErrorResponse(UNKNOWN_REQUEST_MESSAGE))
    }

    // single calling
    return Response.json(await handleCalling(request.body, init))
  })

  /**
   * capture introspection request
   */
  router.use((request, next) => {
    if (request.pathname !== '/__introspection__' || request.method?.toLowerCase() !== 'get') {
      return next()
    }

    return Response.json(info.schema)
  })

  if (options.polling) {
    setInterval(polling, options.pollingInterval)
  }

  return router
}

export const Federation = createFederationServices
