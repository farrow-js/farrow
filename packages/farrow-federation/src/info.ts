import type { FormatResult, FormatEntries } from 'farrow-api/dist/toJSON'
import type { FormatType, FormatTypes } from 'farrow-schema/formatter'
import { updateApi, updateType } from './helpers'
import { getIntrospection } from './introspection'

import { ApiService, ApiServices, FederationOptions, ApiEntryMap } from './federation'

export type FederationInfo = {
  schema: FormatResult
  entryMap: ApiEntryMap
}

export const getFederationInfo = async (
  services: ApiServices,
  options: Required<FederationOptions>,
): Promise<FederationInfo> => {
  const schema: FormatResult = {
    protocol: 'Farrow-API',
    types: {},
    entries: {
      type: 'Entries',
      entries: {},
    },
  }
  const entryMap: ApiEntryMap = new Map()

  let typeBaseId = 0
  const addType = (typeId: number, type: FormatType) => {
    schema.types[typeBaseId + typeId] = updateType(type, typeBaseId)
  }

  const concatTypes = (types: FormatTypes) => {
    let max = 0

    for (const typeId in types) {
      const type = types[typeId]
      const tid = Number(typeId)

      if (tid > max) {
        max = tid
      }

      addType(tid, type)
    }

    return max
  }

  const updateEntries = (entries: FormatEntries, apiService: ApiService, path: string[]): FormatEntries => {
    const newEntries: FormatEntries = {
      type: 'Entries',
      entries: {},
    }

    for (const namespace in entries.entries) {
      const item = entries.entries[namespace]

      if (item.type === 'Entries') {
        newEntries.entries[namespace] = updateEntries(item, apiService, [...path, namespace])
      } else {
        newEntries.entries[namespace] = updateApi(item, typeBaseId)
        entryMap.set(apiService.namespace, apiService.url)
      }
    }

    return newEntries
  }

  const concat = (nextSchema: FormatResult, apiService: ApiService) => {
    const max = concatTypes(nextSchema.types)

    if (schema.entries.entries[apiService.namespace]) {
      const message = `[Federation] Can't merge the service at ${apiService.url} because the namespace: ${apiService.namespace} has been used`
      if (options.strict) {
        throw new Error(message)
      } else {
        console.error(message)
      }
    } else {
      const entries = updateEntries(nextSchema.entries, apiService, [])
      schema.entries.entries[apiService.namespace] = entries
    }

    typeBaseId += max
  }

  for (const service of services) {
    const result = await getIntrospection(service.url, options.fetch)

    if (result.isOk) {
      concat(result.value, service)
    } else {
      const message = `[Federation]: Can't connect to the service at ${service.url} because ${result.value}`
      if (options.strict) {
        throw new Error(message)
      } else {
        console.error(message)
      }
    }
  }

  return {
    schema,
    entryMap,
  }
}
