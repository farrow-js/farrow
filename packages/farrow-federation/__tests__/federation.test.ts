import { Http } from 'farrow-http'
import { service as foo } from './fixtures/foo'
import { service as bar } from './fixtures/bar'
import { services as baz } from './fixtures/baz'
import { createFederationServices } from '../src'
import supertest, { Response } from 'supertest'
import { IncomingMessage, Server, ServerResponse } from 'http'

const fakeFetch = (server: Server<typeof IncomingMessage, typeof ServerResponse>): typeof fetch =>
  (async (input, init) => {
    const getRes = () => {
      if (init?.method?.toLowerCase() === 'post') {
        return supertest(server)
          .post(input as string)
          .send(JSON.parse(init?.body as any))
      }
      return supertest(server).get(input as string)
    }
    const res = await getRes()
    return {
      json() {
        return JSON.parse(res.text)
      },
    }
  }) as typeof fetch

describe('farrow-federation', () => {
  const http = Http()

  http.route('/foo').use(foo)
  http.route('/bar').use(bar)
  http.use(baz)

  it('Introspection', async () => {
    const federationService = await createFederationServices(
      [
        {
          url: '/foo',
          namespace: 'foo',
        },
        {
          url: '/bar',
          namespace: 'bar',
        },
        {
          url: '/greet',
          namespace: 'greet',
        },
        {
          url: '/todo',
          namespace: 'todo',
        },
      ],
      {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        fetch: fakeFetch(http.server()),
      },
    )

    const server = Http()

    server.use(federationService)
    const res = await supertest(server.handle).get('/__introspection__')

    expect(res.statusCode).toBe(200)
    expect(res.body.protocol).toBe('Farrow-API')
    expect(res.body).toMatchSnapshot()
  })

  it('pass through', async () => {
    const federationService = await createFederationServices(
      [
        {
          url: '/foo',
          namespace: 'foo',
        },
        {
          url: '/bar',
          namespace: 'bar',
        },
        {
          url: '/greet',
          namespace: 'greet',
        },
        {
          url: '/todo',
          namespace: 'todo',
        },
      ],
      {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        fetch: fakeFetch(http.server()),
      },
    )

    const server = Http()

    server.use(federationService)
    const res = await supertest(server.handle)
      .post('/')
      .send({
        type: 'Single',
        path: ['foo', 'getTodos'],
        input: { id: 'foo' },
      })

    console.log('0', res.body)

    expect(res.body.type).toBe('ApiSingleSuccessResponse')
    expect(typeof res.body.output).toBe('object')
    expect(Array.isArray(res.body.output.todos)).toBeTruthy()
    expect(res.body.output.todos.length).toBe(3)
  })

  it('mutiple', async () => {
    const federationService = await createFederationServices(
      [
        {
          url: '/foo',
          namespace: 'foo',
        },
        {
          url: '/bar',
          namespace: 'bar',
        },
        {
          url: '/greet',
          namespace: 'greet',
        },
        {
          url: '/todo',
          namespace: 'todo',
        },
      ],
      {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        fetch: fakeFetch(http.server()),
      },
    )

    const server = Http()

    server.use(federationService)
    const res = await supertest(server.handle)
      .post('/')
      .send({
        type: 'Single',
        path: ['bar', 'getTodos'],
        input: { id: 'foo' },
      })

    console.log('1', res.body)

    expect(res.body.type).toBe('ApiSingleSuccessResponse')
    expect(typeof res.body.output).toBe('object')
    expect(Array.isArray(res.body.output.todos)).toBeTruthy()
    expect(res.body.output.todos.length).toBe(3)
  })

  it('nest', async () => {
    const federationService = await createFederationServices(
      [
        {
          url: '/foo',
          namespace: 'foo',
        },
        {
          url: '/bar',
          namespace: 'bar',
        },
        {
          url: '/greet',
          namespace: 'greet',
        },
        {
          url: '/todo',
          namespace: 'todo',
        },
      ],
      {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        fetch: fakeFetch(http.server()),
      },
    )

    const server = Http()

    server.use(federationService)
    const res = await supertest(server.handle)
      .post('/')
      .send({
        type: 'Single',
        path: ['greet', 'greet'],
        input: { name: 'farrow' },
      })

    console.log('2', res.body)

    expect(res.body.type).toBe('ApiSingleSuccessResponse')
    expect(typeof res.body.output).toBe('object')
    expect(res.body.output.greet).toBe('Hello farrow!')
  })
})
