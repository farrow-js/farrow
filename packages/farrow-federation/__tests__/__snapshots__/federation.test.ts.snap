// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`farrow-federation Introspection 1`] = `
{
  "entries": {
    "entries": {
      "bar": {
        "entries": {
          "getTodos": {
            "input": {
              "$ref": "#/types/9",
              "typeId": 9,
            },
            "output": {
              "$ref": "#/types/13",
              "typeId": 13,
            },
            "type": "Api",
          },
        },
        "type": "Entries",
      },
      "foo": {
        "entries": {
          "getTodos": {
            "input": {
              "$ref": "#/types/0",
              "typeId": 0,
            },
            "output": {
              "$ref": "#/types/4",
              "typeId": 4,
            },
            "type": "Api",
          },
        },
        "type": "Entries",
      },
      "greet": {
        "entries": {
          "greet": {
            "description": "Greeting",
            "input": {
              "$ref": "#/types/18",
              "typeId": 18,
            },
            "output": {
              "$ref": "#/types/20",
              "typeId": 20,
            },
            "type": "Api",
          },
        },
        "type": "Entries",
      },
      "todo": {
        "entries": {
          "addTodo": {
            "description": "add todo",
            "input": {
              "$ref": "#/types/27",
              "typeId": 27,
            },
            "output": {
              "$ref": "#/types/30",
              "typeId": 30,
            },
            "type": "Api",
          },
          "getTodos": {
            "description": "get todos",
            "input": {
              "$ref": "#/types/20",
              "typeId": 20,
            },
            "output": {
              "$ref": "#/types/21",
              "typeId": 21,
            },
            "type": "Api",
          },
          "new": {
            "entries": {
              "clearCompleted": {
                "description": "clear completed",
                "input": {
                  "$ref": "#/types/45",
                  "typeId": 45,
                },
                "output": {
                  "$ref": "#/types/46",
                  "typeId": 46,
                },
                "type": "Api",
              },
              "removeTodo": {
                "deprecated": "test",
                "description": "remove todo",
                "input": {
                  "$ref": "#/types/33",
                  "typeId": 33,
                },
                "output": {
                  "$ref": "#/types/36",
                  "typeId": 36,
                },
                "type": "Api",
              },
              "updateTodo": {
                "description": "update todo",
                "input": {
                  "$ref": "#/types/39",
                  "typeId": 39,
                },
                "output": {
                  "$ref": "#/types/43",
                  "typeId": 43,
                },
                "type": "Api",
              },
            },
            "type": "Entries",
          },
        },
        "type": "Entries",
      },
    },
    "type": "Entries",
  },
  "protocol": "Farrow-API",
  "types": {
    "0": {
      "fields": {
        "id": {
          "$ref": "#/types/1",
          "typeId": 1,
        },
      },
      "type": "Struct",
    },
    "1": {
      "type": "Scalar",
      "valueName": "String",
      "valueType": "string",
    },
    "10": {
      "type": "Scalar",
      "valueName": "String",
      "valueType": "string",
    },
    "11": {
      "fields": {
        "todos": {
          "$ref": "#/types/16",
          "typeId": 16,
        },
        "type": {
          "$ref": "#/types/14",
          "typeId": 14,
        },
      },
      "type": "Struct",
    },
    "12": {
      "fields": {
        "message": {
          "$ref": "#/types/10",
          "typeId": 10,
        },
        "type": {
          "$ref": "#/types/17",
          "typeId": 17,
        },
      },
      "type": "Struct",
    },
    "13": {
      "itemTypes": [
        {
          "$ref": "#/types/11",
          "typeId": 11,
        },
        {
          "$ref": "#/types/12",
          "typeId": 12,
        },
      ],
      "type": "Union",
    },
    "14": {
      "type": "Literal",
      "value": "GetTodosSuccess",
    },
    "15": {
      "fields": {
        "content": {
          "$ref": "#/types/10",
          "typeId": 10,
        },
        "createTime": {
          "$ref": "#/types/18",
          "typeId": 18,
        },
      },
      "type": "Struct",
    },
    "16": {
      "$ref": "#/types/15",
      "itemTypeId": 15,
      "type": "List",
    },
    "17": {
      "type": "Literal",
      "value": "UnknownID",
    },
    "18": {
      "fields": {
        "name": {
          "$ref": "#/types/19",
          "description": "The name for greeting",
          "typeId": 19,
        },
      },
      "name": "GreetInput",
      "type": "Object",
    },
    "19": {
      "type": "Scalar",
      "valueName": "String",
      "valueType": "string",
    },
    "2": {
      "fields": {
        "todos": {
          "$ref": "#/types/7",
          "typeId": 7,
        },
        "type": {
          "$ref": "#/types/5",
          "typeId": 5,
        },
      },
      "type": "Struct",
    },
    "20": {
      "fields": {},
      "type": "Struct",
    },
    "21": {
      "fields": {
        "todos": {
          "$ref": "#/types/23",
          "description": "all todos",
          "typeId": 23,
        },
      },
      "type": "Struct",
    },
    "22": {
      "fields": {
        "completed": {
          "$ref": "#/types/26",
          "description": "todo status",
          "typeId": 26,
        },
        "content": {
          "$ref": "#/types/25",
          "description": "todo content",
          "typeId": 25,
        },
        "id": {
          "$ref": "#/types/24",
          "description": "todo id",
          "typeId": 24,
        },
      },
      "name": "Todo",
      "type": "Object",
    },
    "23": {
      "$ref": "#/types/22",
      "itemTypeId": 22,
      "type": "List",
    },
    "24": {
      "type": "Scalar",
      "valueName": "Int",
      "valueType": "number",
    },
    "25": {
      "type": "Scalar",
      "valueName": "String",
      "valueType": "string",
    },
    "26": {
      "type": "Scalar",
      "valueName": "Boolean",
      "valueType": "boolean",
    },
    "27": {
      "fields": {
        "content": {
          "$ref": "#/types/25",
          "description": "todo content to add",
          "typeId": 25,
        },
      },
      "name": "AddTodoInput",
      "type": "Object",
    },
    "28": {
      "fields": {
        "message": {
          "$ref": "#/types/25",
          "typeId": 25,
        },
        "type": {
          "$ref": "#/types/31",
          "typeId": 31,
        },
      },
      "name": "InvalidAddTodoInput",
      "type": "Object",
    },
    "29": {
      "fields": {
        "todo": {
          "$ref": "#/types/22",
          "description": "a new todo",
          "typeId": 22,
        },
        "type": {
          "$ref": "#/types/32",
          "typeId": 32,
        },
      },
      "name": "AddTodoSuccess",
      "type": "Object",
    },
    "3": {
      "fields": {
        "message": {
          "$ref": "#/types/1",
          "typeId": 1,
        },
        "type": {
          "$ref": "#/types/8",
          "typeId": 8,
        },
      },
      "type": "Struct",
    },
    "30": {
      "itemTypes": [
        {
          "$ref": "#/types/28",
          "typeId": 28,
        },
        {
          "$ref": "#/types/29",
          "typeId": 29,
        },
      ],
      "name": "AddTodoOutput",
      "type": "Union",
    },
    "31": {
      "type": "Literal",
      "value": "InvalidAddTodoInput",
    },
    "32": {
      "type": "Literal",
      "value": "AddTodoSuccess",
    },
    "33": {
      "fields": {
        "todoId": {
          "$ref": "#/types/24",
          "description": "todo id wait for removing",
          "typeId": 24,
        },
      },
      "name": "RemoveTodoInput",
      "type": "Object",
    },
    "34": {
      "fields": {
        "todoId": {
          "$ref": "#/types/24",
          "description": "invalid todo id",
          "typeId": 24,
        },
        "type": {
          "$ref": "#/types/37",
          "typeId": 37,
        },
      },
      "name": "TodoIdNotFound",
      "type": "Object",
    },
    "35": {
      "fields": {
        "todoId": {
          "$ref": "#/types/24",
          "description": "todo id that removed",
          "typeId": 24,
        },
        "todos": {
          "$ref": "#/types/23",
          "description": "current todos",
          "typeId": 23,
        },
        "type": {
          "$ref": "#/types/38",
          "typeId": 38,
        },
      },
      "name": "RemoveTodoSuccess",
      "type": "Object",
    },
    "36": {
      "itemTypes": [
        {
          "$ref": "#/types/34",
          "typeId": 34,
        },
        {
          "$ref": "#/types/35",
          "typeId": 35,
        },
      ],
      "type": "Union",
    },
    "37": {
      "type": "Literal",
      "value": "TodoIdNotFound",
    },
    "38": {
      "type": "Literal",
      "value": "RemoveTodoSuccess",
    },
    "39": {
      "fields": {
        "completed": {
          "$ref": "#/types/41",
          "description": "new todo status",
          "typeId": 41,
        },
        "content": {
          "$ref": "#/types/40",
          "description": "new todo content",
          "typeId": 40,
        },
        "todoId": {
          "$ref": "#/types/24",
          "description": "todo id wait for update",
          "typeId": 24,
        },
      },
      "name": "UpdateTodoInput",
      "type": "Object",
    },
    "4": {
      "itemTypes": [
        {
          "$ref": "#/types/2",
          "typeId": 2,
        },
        {
          "$ref": "#/types/3",
          "typeId": 3,
        },
      ],
      "type": "Union",
    },
    "40": {
      "$ref": "#/types/25",
      "itemTypeId": 25,
      "type": "Nullable",
    },
    "41": {
      "$ref": "#/types/26",
      "itemTypeId": 26,
      "type": "Nullable",
    },
    "42": {
      "fields": {
        "todoId": {
          "$ref": "#/types/24",
          "description": "todo id that updated",
          "typeId": 24,
        },
        "todos": {
          "$ref": "#/types/23",
          "description": "current todos",
          "typeId": 23,
        },
        "type": {
          "$ref": "#/types/44",
          "typeId": 44,
        },
      },
      "name": "UpdateTodoSuccess",
      "type": "Object",
    },
    "43": {
      "itemTypes": [
        {
          "$ref": "#/types/34",
          "typeId": 34,
        },
        {
          "$ref": "#/types/42",
          "typeId": 42,
        },
      ],
      "name": "UpdateTodoOutput",
      "type": "Union",
    },
    "44": {
      "type": "Literal",
      "value": "UpdateTodoSuccess",
    },
    "45": {
      "fields": {},
      "type": "Struct",
    },
    "46": {
      "fields": {
        "todos": {
          "$ref": "#/types/23",
          "description": "current todos",
          "typeId": 23,
        },
      },
      "type": "Struct",
    },
    "5": {
      "type": "Literal",
      "value": "GetTodosSuccess",
    },
    "6": {
      "fields": {
        "content": {
          "$ref": "#/types/1",
          "typeId": 1,
        },
        "createTime": {
          "$ref": "#/types/9",
          "typeId": 9,
        },
      },
      "type": "Struct",
    },
    "7": {
      "$ref": "#/types/6",
      "itemTypeId": 6,
      "type": "List",
    },
    "8": {
      "type": "Literal",
      "value": "UnknownID",
    },
    "9": {
      "fields": {
        "id": {
          "$ref": "#/types/10",
          "typeId": 10,
        },
      },
      "type": "Struct",
    },
  },
}
`;
